### PR TITLE
gh-119105: difflib: improve recursion for degenerate cases

### LINKED ISSUE #119105: difflib.py Differ.compare is too slow [for degenerate cases]
# Bug report

### Bug description:

```python
import difflib

a = ["0123456789\n"] * 1_000
b = ["01234a56789\n"] * 1_000
list(difflib.Differ().compare(a, b))  # very slow and will probably hit max recursion depth
```
The case is pathological in the sense of many lines with the same exact diff / ratio.

The issue is that in the current implementation `_fancy_replace` will take the first pair of lines (with the same ratio) as a split point and will call itself recursively for all lines starting at 2, then 3, 4, etc. This repeats `1_000` times resulting in a massive recursion depth and `O(N^3)` complexity scaling.

For an average random case it should split anywhere in the range of  `1_000` with a complexity scaling of `O(N^2 log N)`.

I personally encountered this in diffing csv files where one of the files has a column added which, apparently, results in all-same ratios for every line in the file.

### Proposal

Fixing this is not so hard by adding some heuristics (WIP) https://github.com/pulkin/cpython/commit/31e1ed03cf05bdbc9b4695c8bb680e65963f9bff

The idea is very straightforward: while doing the `_fancy_replace` magic, if you see many diffs with the same exact ratio, pick the one closest to the middle of chunks rather than the first one (which can be the worst possible choice). This is done by promoting the ratios of those pairs of lines that are closer to the middle of the chunks.

The `_drag_to_center` function turns line number into the weight added to the ratio (twice: for a and for b). The weight is zero for both ends of chunks and maximal in the middle (quadratic poly was chosen for simplicity). The magnitude of the weight "`_gravity`" is small enough to only affect ratios that are exactly the same: it relies on the assumption that we probably have less than 500k symbols in the line such that the steps in the `ratio` are greater than 1e-6. If this assumption fails some diffs may become different (not necessarily worse).

Performance impact for non-pathological cases is probably minimal.

### CPython versions tested on:

3.9, 3.12

### Operating systems tested on:

Linux

### Linked PRs
* gh-119131
* gh-119376
* gh-119492

### PR DESCRIPTION
**Pull request by pulkin**
Used this as a benchmark:

```python
a = ["0123456789\n"] * 1_000
b = ["01234a56789\n"] * 1_000

from difflib import Differ
import timeit

print(timeit.timeit("list(Differ().compare(a, b))", number=1, globals=globals()))
# this exceeds recursion depth on my stock python distribution.
# once `difflib` is patched, it runs in approximately `1s`
```

* Issue: gh-119105

### CODE REVIEW

**Comment by tim-one** [comment]
You may have noticed that it's bugging you for a NEWS entry. Easiest is to click "Details" on the failing "bedevere/news" check, That will bring up a web page that asks you for the GitHub issue number, PR number, which part of the distribution is affected ("library", here), and type some text explaining the change at a high (user-oriented) level. When that's done, submit the page, and it will magically generate a new file with a crazy name,  and commit it to this branch in the `Misc/NEWS.d` directory.

File: Lib/difflib.py [hunk 1]
```diff
@@ -911,33 +911,48 @@ def _fancy_replace(self, a, alo, ahi, b, blo, bhi):
 
         # don't synch up unless the lines have a similarity score of at
         # least cutoff; best_ratio tracks the best score seen so far
-        best_ratio, cutoff = 0.74, 0.75
+        best_ratio, cutoff = (0.74, 0), 0.75
```
Comment (tim-one) [inline Lib/difflib.py:921]:
```suggestion
        # best_ratio is a tuple storing the best .ratio() seen so far, and
        # a measure of how far the indices are from their index range
        # midpoints. The latter is used to resolve ratio ties. Favoring
        # indices near the midpoints tends to cut the ranges in half. Else,
        # if there are many pairs with the best ratio, recursion can grow
        # very deep, and runtime becomes cubic. See:
        # https://github.com/python/cpython/issues/119105
        best_ratio, cutoff = (0.74, 0), 0.75
```

File: Misc/NEWS.d/next/Library/2024-05-19-12-25-36.gh-issue-119105.VcR4ig.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+``difflib.Differ`` is faster for some class of diffs
```
Comment (tim-one) [inline Misc/NEWS.d/next/Library/2024-05-19-12-25-36.gh-issue-119105.VcR4ig.rst]:
```suggestion
``difflib.Differ`` is much faster for some cases of diffs where many pairs of lines are equally similar.
```

**Review by tim-one — APPROVED** [review]
Looks good to go! It's been a pleasure working with you - I hope you stick around :smile:.

**Comment by tim-one** [comment]
BTW, I'm not sure whether it will let you "squash and merge" and the change. If not, let me know and I'll do the commit.

**Comment by pulkin** [comment]
Just letting you know that I am fine with the merge. I am not allowed to merge.

Thanks for the guidance and discussion btw.

**Comment by tim-one** [comment]
All done! You're an official Python Committer™ now. Use your power only for good 😉

**Comment by tim-one** [comment]
I believe the buildbot failures are spurious "false positives" for this PR. One is failing a threading test, and the other sqlite3. `difflib` has no plausible connection to either. When a PR is "really" at fault, some dozens of these buildbot failures are generated.

**Pull request merged**